eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs

```diff
@@ -321,30 +321,49 @@ delegationTransition = do
           }
-    DCertMir (MIRCert targetPot (StakeAddressesMIR credCoinMap)) -> do
-      if HardForks.allowMIRTransfer pp
-        then do
-          sp <- liftSTS $ asks stabilityWindow
-          ei <- liftSTS $ asks epochInfoPure
-          EpochNo currEpoch <- liftSTS $ epochInfoEpoch ei slot
-          let newEpoch = EpochNo (currEpoch + 1)
-          tellEvent (NewEpoch newEpoch)
-          firstSlot <- liftSTS $ epochInfoFirst ei newEpoch
-          let tooLate = firstSlot *- Duration sp
-          slot
-            < tooLate
-            ?! MIRCertificateTooLateinEpochDELEG slot tooLate
-
-          let (potAmount, delta, instantaneousRewards) =
-                case targetPot of
-                  ReservesMIR -> (asReserves acnt, deltaReserves $ dsIRewards ds, iRReserves $ dsIRewards ds)
-                  TreasuryMIR -> (asTreasury acnt, deltaTreasury $ dsIRewards ds, iRTreasury $ dsIRewards ds)
-              credCoinMap' = Map.map (\(DeltaCoin x) -> Coin x) credCoinMap
-              combinedMap = Map.unionWith (<>) credCoinMap' instantaneousRewards
-              requiredForRewards = fold combinedMap
-              available = potAmount `addDeltaCoin` delta
+    DCertMir (MIRCert targetPot mirTarget) ->
+      sp <- liftSTS $ asks stabilityWindow
+      ei <- liftSTS $ asks epochInfoPure
+      EpochNo currEpoch <- liftSTS $ epochInfoEpoch ei slot
+      let newEpoch = EpochNo (currEpoch + 1)
+      tellEvent (NewEpoch newEpoch)
+      firstSlot <- liftSTS $ epochInfoFirst ei newEpoch
+      let tooLate = firstSlot *- Duration sp
+      slot
+        < tooLate
+        ?! MIRCertificateTooLateinEpochDELEG slot tooLate
+
+      case mirTarget of
+        (StakeAddressesMIR credCoinMap) -> do
+
+          (potAmount, delta, combinedMap) <- if HardForks.allowMIRTransfer pp
+                then do
+                  let (potAmount, delta, instantaneousRewards) =
+                        case targetPot of
+                          ReservesMIR -> (asReserves acnt, deltaReserves $ dsIRewards ds, iRReserves $ dsIRewards ds)
+                          TreasuryMIR -> (asTreasury acnt, deltaTreasury $ dsIRewards ds, iRTreasury $ dsIRewards ds)
+
+                  let credCoinMap' = Map.map (\(DeltaCoin x) -> Coin x) credCoinMap
+                      combinedMap = Map.unionWith (<>) credCoinMap' instantaneousRewards
+
+                  all (>= mempty) combinedMap ?! MIRProducesNegativeUpdate
+
+                  pure (potAmount, delta, combinedMap)
+                else do
+                  let (potAmount, instantaneousRewards) =
+                        case targetPot of
+                          ReservesMIR -> (asReserves acnt, iRReserves $ dsIRewards ds)
+                          TreasuryMIR -> (asTreasury acnt, iRTreasury $ dsIRewards ds)
+
+                  let credCoinMap' = Map.map (\(DeltaCoin x) -> Coin x) credCoinMap
+                      combinedMap = Map.union credCoinMap' instantaneousRewards
```

Comment:
Here is some more deduplication (also removes a few lines from below, just can't add this as suggestion). Note that `delta` is computed lazily:
```suggestion
          let (potAmount, delta, instantaneousRewards) =
                case targetPot of
                  ReservesMIR -> (asReserves acnt, deltaReserves $ dsIRewards ds, iRReserves $ dsIRewards ds)
                  TreasuryMIR -> (asTreasury acnt, deltaTreasury $ dsIRewards ds, iRTreasury $ dsIRewards ds)
              credCoinMap' = Map.map (\(DeltaCoin x) -> Coin x) credCoinMap

          (requiredForRewards, available) <- if HardForks.allowMIRTransfer pp
            then do
              let combinedMap = Map.unionWith (<>) credCoinMap' instantaneousRewards
              all (>= mempty) combinedMap ?! MIRProducesNegativeUpdate
              pure (fold combinedMap, potAmount `addDeltaCoin` delta)
            else do
              let combinedMap = Map.union credCoinMap' instantaneousRewards
              all (>= mempty) credCoinMap ?! MIRNegativesNotCurrentlyAllowed
              pure (fold combinedMap, potAmount)
                      
```